File: ios/ObjectiveC/Beacon/9-Alert-Strategy/Complete/beaconAlertComplete/SecondWayAlertTimerStrategy.m
//
//  SecondWayAlertTimerStrategy.m
//  beaconAlertComplete
//
//  Created by sarra srairi on 02/08/2016.
//  Copyright © 2016 sarra srairi. All rights reserved.
//

#import "SecondWayAlertTimerStrategy.h"

@implementation SecondWayAlertTimerStrategy
@synthesize timeToCreateAlert;


- (id)initWithMaxTime:(int)_maxTimeBeforeReset
DelayAfterDetectingBeaconToCreateAlert:(int)_timeToCreateAlert
{
     self = [super initWithMaxTime:_maxTimeBeforeReset];
    if (self) {
 
        self.timeToCreateAlert = ((double) _timeToCreateAlert / 1000)+ CFAbsoluteTimeGetCurrent();
        
    }
    return self;
}

-(NSString *) getKey{
    return @"timeAlertStrategy";
}


-(bool)isConditionValid:(ATBeaconAlertParameterProximity *) beaconParameter {
    return ([beaconParameter isInProximityForAction]);
}
 
-(BOOL)isInProximityForAction : (ATBeaconContent *)beaconContent {
    if ([beaconContent poi] == nil){
        return false;
    }
    NSString * rangeString = [beaconContent getRange];
    if (rangeString || [@""isEqualToString:rangeString]){
        if (beaconContent.beacon != nil && CFAbsoluteTimeGetCurrent() >= timeToCreateAlert){
            return YES;
        }
    }
    return false;
}


-(void) updateAlertParameter:(ATBeaconContent *) _beaconContent{
    
    ATBeaconAlertParameterProximity *beaconAlertParameterProximty = (ATBeaconAlertParameterProximity *)_beaconContent.beaconAlertParameter;
    beaconAlertParameterProximty.isInProximityForAction = [self isInProximityForAction:_beaconContent];
    
}
 
-(ATBeaconAlertParameter *) createBeaconAlertParameter {
    return ([[ATBeaconAlertParameterProximity alloc] init]);
}

@end

File: ios/SWIFT/Beacon/9-Alert-Strategy/Complete/AlertStrategy/SecondWayAlertTimerStrategy.swift
//
//  SecondWayAlertTimerStrategy.swift
//  AlertStrategy
//
//  Created by sarra srairi on 11/08/2016.
//  Copyright © 2016 R&D connecthings. All rights reserved.
//
import UIKit
import ATAnalytics
import ATLocationBeacon
import AVFoundation


class SecondWayAlertTimerStrategy: ATBeaconAlertStrategy {
    
    var timeToCreateAlert: Double!
    
    // The output below is limited by 1 KB.
    // Please Sign Up (Free!) to remove this limitation.
    
    
    init(maxTime maxTimeBeforeReset: Int, DelayAfterDetectingBeaconToCreateAlert timeToCreateAlert: Int) {
        super.init(maxTime: (Int32(maxTimeBeforeReset)))
        
        self.timeToCreateAlert = (Double(timeToCreateAlert) / 1000 )
        NSLog("time to create the alert consutuct", self.timeToCreateAlert);
        //+ CFAbsoluteTimeGetCurrent()
    }
    
    override func getKey() -> String {
        return "timeAlertStrategy"
    }
    
    override func isConditionValid(_beaconParameter: ATBeaconAlertParameter!) -> Bool {
        let beaconAlertParameterProximty: ATBeaconAlertParameterProximity = (_beaconParameter  as! ATBeaconAlertParameterProximity)
        return beaconAlertParameterProximty.isInProximityForAction
    }
    
    func isInProximityForAction(beaconContent: ATBeaconContent) -> Bool {
        if beaconContent.poi == nil {
            return false
        }
        
        var rangeString: String!
        rangeString = beaconContent.getRange()
        
        if rangeString.isEmpty  {
            if beaconContent.beacon != nil && CFAbsoluteTimeGetCurrent() >= timeToCreateAlert {
                return true
            }
        }
        return false
    }
    
    override func updateAlertParameter(beaconContent: ATBeaconContent) {
        let beaconAlertParameterProximty: ATBeaconAlertParameterProximity = (beaconContent.beaconAlertParameter as! ATBeaconAlertParameterProximity)
        beaconAlertParameterProximty.isInProximityForAction = self.isInProximityForAction(beaconContent)
    }
    override func createBeaconAlertParameter() -> ATBeaconAlertParameter {
        return (ATBeaconAlertParameter())
    }
}